libs/utils/src/cast_vote_records.ts

```diff
@@ -0,0 +1,459 @@
+}): CVR.CVRContest {
+  const overvoted = vote.length > contest.seats;
+  const undervoted = vote.length < contest.seats;
+
+  const statuses: CVR.ContestStatus[] = [];
+  if (vote.length === 0) {
+    statuses.push(CVR.ContestStatus.NotIndicated);
+  }
+
+  if (undervoted) {
+    statuses.push(CVR.ContestStatus.Undervoted);
+  }
+
+  if (overvoted) {
+    statuses.push(
+      CVR.ContestStatus.Overvoted,
+      CVR.ContestStatus.InvalidatedRules
+    );
+  }
+
+  const writeInCount = vote.reduce(
+    (count, choice) => count + (choice.isWriteIn ? 1 : 0),
+    0
+  );
+
+  // We track the write in counter in order to give the BMD write-ins an
+  // increasing index. For HMPB write-ins, we use the index in the write-in id
+  let writeInCounter = 0;
+
+  return {
+    '@type': 'CVR.CVRContest',
+    ContestId: contest.id,
+    Overvotes: overvoted ? vote.length - contest.seats : undefined,
+    Undervotes: undervoted ? contest.seats - vote.length : undefined,
+    WriteIns: writeInCount > 0 ? writeInCount : undefined,
+    Status: statuses,
+    Selections: contest.seats,
+    CVRContestSelection: vote.map((candidate) => {
+      const { isWriteIn } = candidate;
+      const ContestSelectionId =
+        candidate.isWriteIn && ballotMarkingMode === 'machine'
+          ? `write-in-${writeInCounter}`
+          : candidate.id;
+      const OptionPosition = !candidate.isWriteIn
+        ? getCandidateOptionPosition({ contest, candidate })
+        : contest.candidates.length +
+          (ballotMarkingMode === 'hand'
+            ? getHmpbWriteInCandidateIndex(candidate)
+            : writeInCounter);
+      if (isWriteIn) writeInCounter += 1;
+
+      return {
+        '@type': 'CVR.CVRContestSelection',
+        ContestSelectionId,
+        OptionPosition,
+        Status: overvoted
+          ? [CVR.ContestSelectionStatus.InvalidatedRules]
+          : isWriteIn
+          ? [CVR.ContestSelectionStatus.NeedsAdjudication]
+          : undefined,
+        SelectionPosition: [
+          {
+            '@type': 'CVR.SelectionPosition',
+            HasIndication: CVR.IndicationStatus.Yes,
+            NumberVotes: 1,
+            IsAllocable: overvoted
+              ? CVR.AllocationStatus.No
+              : isWriteIn
+              ? CVR.AllocationStatus.Unknown
+              : CVR.AllocationStatus.Yes,
+            Status: overvoted
+              ? [CVR.PositionStatus.InvalidatedRules]
+              : undefined,
+            CVRWriteIn: isWriteIn
+              ? {
+                  '@type': 'CVR.CVRWriteIn',
+                  Text:
+                    ballotMarkingMode === 'machine'
+                      ? candidate.name
+                      : undefined,
+                  // TODO: Once ballot images are separate file, include filename here for HMPB
+                }
+              : undefined,
+          },
+        ],
+      };
+    }),
+  };
+}
+
+/**
+ * Blah
+ */
+export function buildCVRContestsFromVotes({
```

Comment:
I wanted to put as much as possible into `libs/utils` so I could re-use it for fixture generation.